core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
-          case None =>
-            // this one should be completed by the new coordinator
-            warn(s"no longer the coordinator for transactionalId: $transactionalId")
-        }
-    }
-    txnManager.appendTransactionToLog(transactionalId, newMetadata, logAppendCallback)
-  }
+      result match {
+        case Left(err) =>
+          responseCallback(err)
 
-  private def writeTxnMarkers(markerArgs: WriteTxnMarkerArgs): Unit = {
-    def completionCallback(error: Errors): Unit = {
-      error match {
-        case Errors.NONE =>
-          txnManager.getTransactionState(markerArgs.transactionalId) match {
-            case Some(preparedCommitMetadata) =>
-              val completedState = if (markerArgs.nextState == PrepareCommit) CompleteCommit else CompleteAbort
-              val committedMetadata = new TransactionMetadata(markerArgs.pid,
-                markerArgs.epoch,
-                preparedCommitMetadata.txnTimeoutMs,
-                completedState,
-                preparedCommitMetadata.topicPartitions,
-                preparedCommitMetadata.transactionStartTime,
-                time.milliseconds())
-              preparedCommitMetadata.prepareTransitionTo(completedState)
-
-              def writeCommittedTransactionCallback(error: Errors): Unit = {
-                error match {
-                  case Errors.NONE =>
-                    txnMarkerChannelManager.removeCompleted(txnManager.partitionFor(markerArgs.transactionalId),
-                      markerArgs.pid)
-                  case Errors.NOT_COORDINATOR =>
-                    // this one should be completed by the new coordinator
-                    warn(s"no longer the coordinator for transactionalId: ${markerArgs.transactionalId}")
-                  case _ =>
-                    warn(s"error: $error caught for transactionalId: ${markerArgs.transactionalId} when appending state: $completedState. Retrying.")
-                    // retry until success
-                    txnManager.appendTransactionToLog(markerArgs.transactionalId, committedMetadata, writeCommittedTransactionCallback)
-                }
+        case Right((coordinatorEpoch, newMetadata)) =>
+          def sendTxnMarkersCallback(error: Errors): Unit = {
+            // we can respond to the client immediately and continue to write the txn markers if
+            // the log append was successful
+            responseCallback(error)
+
+            if (error == Errors.NONE) {
+              txnManager.getTransactionState(transactionalId) match {
+                case Some(epochAndMetadata) =>
+                  val metadata = epochAndMetadata.transactionMetadata
+
+                  sendTxnMarkers(transactionalId, coordinatorEpoch, metadata, txnMarkerResult)
+
+                case None =>
+                  // this transactional id no longer exists, maybe the corresponding partition has already been migrated out.
```

Comment:
Could we check for the case explicitly with `txnManager.isCoordinatorFor(transactionalId)`? If the transactionalId did not migrate, then maybe it should be an illegal state.